### DISCUSSION TITLE
Issues when combining login with password and passwordless in the same model.

### DISCUSSION THREAD

**Discussion by JoseMariaSierra** [abevoelker/devise-passwordless#60] [Q&A]
Hello!

I apologize in advance if I'm not understanding the gem's functionality correctly, but I'm encountering issues when using both the native login system of Devise and the passwordless one in the same model.

In my user model, I have the following setup

```ruby
class User < ApplicationRecord

  devise :magic_link_authenticatable, :database_authenticatable, :registerable, :recoverable, :rememberable,
         :validatable, :confirmable, :trackable
         
  def active_for_authentication?
    super && password_auth?
  end

  def inactive_message
    password_auth? ? super : :password_authentication_disabled
  end

  def active_for_magic_link_authentication?
    super && magic_links?
  end

  def magic_link_inactive_message
    magic_links? ? super : :magic_link_authentication_disabled
  end
  
  end
```

That means, I've implemented logic similar to [Disabling password authentication or magic link authentication](https://github.com/abevoelker/devise-passwordless?tab=readme-ov-file#combining-password-and-passwordless-auth-in-the-same-model).

`magic_link` and `password_auth` are two boolean columns used to determine whether a user utilizes one or the other authorization system. For testing purposes, they're set as follows: `magic_link: false`, `password_auth: true`.

With this setup, a user can never log in with a password unless the` active_for_magic_link_authentication? `method returns true. In other words, for some reason, I'm forced to set the `magic_link` column to true to allow a user to log in with a password.

Oddly, the execution always goes through active_for_magic_link_authentication? even if I remove magic_link_authenticatable, and the login is performed with a password.

The only way for a user to use a password, given the gem, is as follows (Or removing the method):

```
def active_for_magic_link_authentication?
    true
  end
```

It's worth noting that I haven't added any additional logic beyond the routes indicated in the documentation.

Could you assist me with this? Is this the expected behavior of the gem?

I thought I could use both systems simultaneously, but I'm not achieving it.

Thank you very much

**Comment by JoseMariaSierra** [comment]
To fix it, what I did was remove the 'active_for_magic_link_authentication?' method and add the use of passwordless to my model conditionally:

```ruby
devise :magic_link_authenticatable if -> { magic_link_enabled? }
```

This way it works correctly.

**Comment by abevoelker** [comment]
:wave: Hey, sorry I didn't get to this before you figured out a solution. As time allows I will try to add your setup to our testing harness and figure out what went wrong. Thanks for coming back to share your solution!

**Reply by JoseMariaSierra** [reply]
No problem, thank you for implementing this gem. It just took a little bit of imagination.

**Comment by collabital** [comment]
I think I've managed to reproduce the issue here:
https://gist.github.com/collabital/fb888e6569e6308efe79aa2a9fccac09

Essentially, it comes down to `database_authenticatable` not having a hook that only runs when it is the winning Warden strategy.
Using `#active_for_authentication?` is too broad, and may affect **any** successful authentication strategy.

In the absence of Devise considering an enhancement to `database_authenticatable`, the fix I would instead suggest is adding support for `#require_magic_link_authentication?`, which would prevent logging in using a strategy other than `MagicLinkAuthenticatable`. Of course further customisation is possible by adding a bespoke Warden initializer.

```
# app/model/user.rb
# Bob must use password authentication. All other users are required to use magic links.
  def first_name_bob?
    self.first_name.downcase == "bob"
  end

  def active_for_magic_link_authentication?
    super && !self.first_name_bob?
  end
  
  # This method needs to be implemented in lib/devise/models/magic_link_authenticatable.rb with an appropriate default
  def require_magic_link_authentication?
    super && !self.first_name_bob?
  end

  def magic_link_inactive_message
    first_name_bob? ? super : :first_name_not_bob
  end

  # This method needs to be implemented in lib/devise/models/magic_link_authenticatable.rb with an appropriate default
  def magic_link_required_message
    first_name_bob? ? super : :first_name_bob
  end
```

Then extending the logic in `lib/devise/hooks/magic_link_authenticatable.rb`:
```
# frozen_string_literal: true

Warden::Manager.after_set_user do |record, warden, options|
  # Deny user magic link access when magic link authentication is disabled
  if record&.respond_to?(:active_for_magic_link_authentication?) && !record.active_for_magic_link_authentication? && warden.winning_strategy.is_a?(Devise::Strategies::MagicLinkAuthenticatable)
    scope = options[:scope]
    warden.logout(scope)
    throw :warden, scope: scope, message: record.magic_link_inactive_message
  end
  # Deny user access when magic link authentication is required but not used
  if record&.respond_to?(:require_magic_link_authentication?) && record.require_magic_link_authentication? && !warden.winning_strategy.is_a?(Devise::Strategies::MagicLinkAuthenticatable)
    scope = options[:scope]
    warden.logout(scope)
    throw :warden, scope: scope, message: record.magic_link_required_message
  end
end
```
